pkg/conversion/generator.go

```diff
@@ -0,0 +1,377 @@
+		indentation = indentation + "\t"
+	}
+	_, err := f.WriteString(indentation)
+	return err
+}
+
+func writeLine(f *os.File, indent int, line string) error {
+	if err := indentLine(f, indent); err != nil {
+		return err
+	}
+	if _, err := f.WriteString(line); err != nil {
+		return err
+	}
+	return nil
+}
+
+func writeHeader(f *os.File, inType, outType string, indent int) error {
+	// FIXME(wojtek-t): Handle packages correctly.
+	format := "func(in *%s, out *newer.%s, s conversion.Scope) error {\n"
+	stmt := fmt.Sprintf(format, inType, outType)
+	return writeLine(f, indent, stmt)
+}
+
+func writeFooter(f *os.File, indent int) error {
+	if err := writeLine(f, indent+1, "return nil\n"); err != nil {
+		return err
+	}
+	if err := writeLine(f, indent, "},\n"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *generator) writeConversionForMap(f *os.File, inField, outField reflect.StructField, indent int) error {
+	ifFormat := "if in.%s != nil {\n"
+	ifStmt := fmt.Sprintf(ifFormat, inField.Name)
+	if err := writeLine(f, indent, ifStmt); err != nil {
+		return err
+	}
+	// FIXME(wojtek-t): Handle packages correctly.
+	makeFormat := "out.%s = make(%s)\n"
+	makeStmt := fmt.Sprintf(makeFormat, outField.Name, outField.Type)
+	if err := writeLine(f, indent+1, makeStmt); err != nil {
+		return err
+	}
+	forFormat := "for key, val := range in.%s {\n"
+	forStmt := fmt.Sprintf(forFormat, inField.Name)
+	if err := writeLine(f, indent+1, forStmt); err != nil {
+		return err
+	}
+	newValue := false
+	if isComplexType(inField.Type.Elem()) || !inField.Type.Elem().ConvertibleTo(outField.Type.Elem()) {
+		newValue = true
+		newFormat := "newVal := %s{}\n"
+		// FIXME(wojtek-t): Handle packages correctly.
+		newStmt := fmt.Sprintf(newFormat, outField.Type.Elem())
+		if err := writeLine(f, indent+2, newStmt); err != nil {
+			return err
+		}
+		convertStmt := "if err := s.Convert(&val, &newVal, 0); err != nil {\n"
+		if err := writeLine(f, indent+2, convertStmt); err != nil {
+			return err
+		}
+		if err := writeLine(f, indent+3, "return err\n"); err != nil {
+			return err
+		}
+		if err := writeLine(f, indent+2, "}\n"); err != nil {
+			return err
+		}
+	}
+	if inField.Type.Key().ConvertibleTo(outField.Type.Key()) {
+		value := "val"
+		if newValue {
+			value = "newVal"
+		}
+		assignStmt := ""
+		if inField.Type.Key().AssignableTo(outField.Type.Key()) {
+			assignStmt = fmt.Sprintf("out.%s[key] = %s\n", outField.Name, value)
+		} else {
+			// FIXME(wojtek-t): Handle packages correctly.
+			assignStmt = fmt.Sprintf("out.%s[%s(key)] = %s\n", outField.Name, outField.Type.Key(), value)
+		}
+		if err := writeLine(f, indent+2, assignStmt); err != nil {
+			return err
+		}
+	}
+	if err := writeLine(f, indent+1, "}\n"); err != nil {
+		return err
+	}
+	if err := writeLine(f, indent, "}\n"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *generator) writeConversionForSlice(f *os.File, inField, outField reflect.StructField, indent int) error {
```

Comment:
golang comment: use io.Reader/io.Writer, not os.File as parameters-- helps with testing among other things.